### ISSUE TITLE
Upstream the Signal stuff to SwiftGodot 

### ISSUE THREAD

**Issue by PadraigK** [PadraigK/SGHelpers#1]
We talked about this in the SwiftGodot slack. 

I think we could do a macro that could be invoked like this:
```
@Godot class Player: Node {
  @Signal("lives_changed", [Argument(name: "count", type: Int.self)])
} 
```
and it would expand to
```
class Player: Node {
  static let livesChanged = SignalWith1Argument<Int>("lives_changed", argumentName: "count" classType: Player.self)

  <init stuff that registers the signal along with any @Callable>
}
```

The macro's role here would be:
1. include the registration in the class init
2. hide the clumsy "With1Argument" variadic-generics work around (we need macOS 14 / Sonoma to use the new stuff this way) 
3. automatically determine the classType argument 

- [x] Build the macro
- [x] See if I can improve propInfo instead of always sending `.object` 
- [x] Auto-generate VariantConvertible conformance

**Comment by PadraigK** [comment]
@matt-is-confused also suggested that we ought to get the other `PropInfo` properties right so that Godot editor can be used to connect a Swift-defined + emitted signal to a GDScript on another node.

**Comment by PadraigK** [comment]
Some notes after looking at the current state of the editor with the SGHelper Signal stuff as it is:

1. Signals that are referenced in `_ready` methods are registered and visible, so that's good — once the macro is including them in the class init, they should all show here regardless of whether they're referenced somewhere that the editor triggers right now.

<img width="202" alt="Screenshot 2023-10-18 at 10 55 31 AM" src="https://github.com/PadraigK/SGHelpers/assets/466150/35dc11f0-a8d3-4853-be0e-fdc9f265432b">

Here, `lives_changed` is one I defined in Swift.

2. As we expected, the argument descriptions are all wrong: `lives_changed(Optional(Swift.Int): Player)` 
  * the argument name is coming in as the type name, but it should be at least something like "arg1" or better yet, something programmer supplied.
  * the argument type is the class type that the signal is defined on, but it should be `Int`, or GType.int even. 

3. It is possible to connect this signal to another node, but the same issues with the PropInfo mean that the GDScript that is generated is invalid. 

<img width="706" alt="Screenshot 2023-10-18 at 10 59 10 AM" src="https://github.com/PadraigK/SGHelpers/assets/466150/132fdf3e-dd8d-42bb-a260-133e463441b4">

It can be done!

<img width="1051" alt="Screenshot 2023-10-18 at 11 00 27 AM" src="https://github.com/PadraigK/SGHelpers/assets/466150/698b64be-6d06-4208-a0b2-3c2af40c0402">

But it doesn't work right now.

----

In conclusion:
1. We can definitely declare signals in swift and have them exposed to the editor
2. We can connect them up in the editor, but we need to fix the info we're sending to PropInfo

**Comment by PadraigK** [comment]
I think I will attempt to fix the PropInfo bits _here_ before upstreaming anything.
